parlai/core/worlds.py

```diff
@@ -975,6 +975,209 @@ def shutdown(self):
+                'World must be a DialogPartnerWorld or a '
+                'MultiWorld of DialogPartnerWorld'
+            )
+
+        if len(world.get_agents()) != 2:
+            raise AssertionError(
+                "Dynamic batch only works in a fixed dialog world with two agents."
+            )
+
+        if not hasattr(world.get_agents()[1], 'batch_act'):
+            raise TypeError("Model agent doesn't have batch_act.")
+
+        self.truncate = opt.get('text_truncate', None) or opt.get('truncate', None)
+        self.l_truncate = opt.get('label_truncate', None) or opt.get('truncate', None)
+        if self.truncate is None or self.truncate < 0:
+            raise ValueError(
+                'You must use --text-truncate or --truncate in order to use '
+                'dynamic batching.'
+            )
+
+        # size of the buffer we will use to find worlds
+        self._BUFFER_SIZE = 1021  # chosen as a prime number
+
+        # TODO: check to ensure the agent has self_observe
+        shared = world.share()
+        self.world = world
+        # TODO: maybe generalize this
+        self.max_size = (self.l_truncate + self.truncate) * opt['batchsize']
+        self.rng = random.Random(4)
+
+        # buffer worlds
+        self.worlds = [
+            shared['world_class'](opt, None, shared) for _ in range(self._BUFFER_SIZE)
+        ]
+
+        self.reset()
+
+    def reset(self):
+        super().reset()
+        self._obs = [None for _ in range(self._BUFFER_SIZE)]
+        self._scores = [None for _ in range(self._BUFFER_SIZE)]
+
+        self.number_parleys = 0
+        self.total_exs = 0
+        self.world.reset()
+        self.rng = random.Random(4)
+        for w in self.worlds:
+            w.reset()
+
+    def reset_metrics(self):
+        super().reset_metrics()
+        self.world.reset_metrics()
+        for w in self.worlds:
+            w.reset_metrics()
+
+    def epoch_done(self):
+        return (
+            self.world.epoch_done()
+            or all(w.epoch_done() for w in self.worlds)
+            and all(s is None for s in self._scores)
+        )
+
+    def num_examples(self):
+        return self.world.num_examples()
+
+    def num_episodes(self):
+        return self.world.num_episodes()
+
+    def _ceil(self, n, r=4):
+        """
+        Round to the nearest multiple of 4.
+
+        TensorCores only work when a tensor is a multiple of 4 in all dimensions. This
+        means all examples cost is related to their nearest multiple of 4.
+        """
+        # round up to 4, all things are equal
+        return ((n + r - 1) // r) * r
+
+    def _score(self, obs):
+        if 'text_vec' in obs:
+            # note that all examples have a cost that is based on their
+            # nearest multiple of 4. We can therefore mix-and-match
+            # anything with the same cost for increased stochasticity,
+            # while not really wasting much padding.
+            return tuple(
```

Comment:
how does this play with label candidates for ranking models?

Reply:
i.e. even if we train with batch candidates, we typically evaluate with inline label candidates, so the memory usage woudl go way up

Reply:
Probably very poorly. Worth testing

Reply:
This has been tested.